File: src/cpp/src/chat_history.cpp
// Copyright (C) 2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

#include "openvino/genai/chat_history.hpp"

namespace ov {
namespace genai {

ChatHistory::ChatHistory() = default;

ChatHistory::ChatHistory(const JsonContainer& messages) : m_messages(messages) {
    if (!m_messages.is_array()) {
        OPENVINO_THROW("Chat history must be initialized with a JSON array.");
    }
}
ChatHistory::ChatHistory(const std::vector<ov::AnyMap>& messages) :
    m_messages(JsonContainer::array()) {
    for (const auto& message : messages) {
        m_messages.push_back(JsonContainer(message));
    }
}

ChatHistory::ChatHistory(std::initializer_list<std::initializer_list<std::pair<std::string, ov::Any>>> messages) :
    m_messages(JsonContainer::array()) {
    for (const auto& message : messages) {
        m_messages.push_back(JsonContainer(message));
    }
}

ChatHistory::~ChatHistory() = default;

ChatHistory& ChatHistory::push_back(const JsonContainer& message) {
    m_messages.push_back(message);
    return *this;
}

ChatHistory& ChatHistory::push_back(const ov::AnyMap& message) {
    m_messages.push_back(JsonContainer(message));
    return *this;
}

ChatHistory& ChatHistory::push_back(std::initializer_list<std::pair<std::string, ov::Any>> message) {
    m_messages.push_back(JsonContainer(message));
    return *this;
}

void ChatHistory::pop_back() {
    if (m_messages.empty()) {
        OPENVINO_THROW("Cannot pop_back from an empty chat history.");
    }
    m_messages.erase(m_messages.size() - 1);
}

const JsonContainer& ChatHistory::get_messages() const {
    return m_messages;
}

JsonContainer& ChatHistory::get_messages() {
    return m_messages;
}

JsonContainer ChatHistory::operator[](size_t index) const {
    if (index >= m_messages.size()) {
        OPENVINO_THROW("Index ", index, " is out of bounds for chat history of size ", m_messages.size());
    }
    return m_messages[index];
}

JsonContainer ChatHistory::operator[](int index) const {
    return operator[](size_t(index));
}

JsonContainer ChatHistory::first() const {
    if (m_messages.empty()) {
        OPENVINO_THROW("Cannot access first message of an empty chat history.");
    }
    return m_messages[0];
}

JsonContainer ChatHistory::last() const {
    if (m_messages.empty()) {
        OPENVINO_THROW("Cannot access last message of an empty chat history.");
    }
    return m_messages[m_messages.size() - 1];
}

void ChatHistory::clear() {
    m_messages.clear();
}

size_t ChatHistory::size() const {
    return m_messages.size();
}

bool ChatHistory::empty() const {
    return m_messages.empty();
}

ChatHistory& ChatHistory::set_tools(const JsonContainer& tools) {
    if (!tools.is_array()) {
        OPENVINO_THROW("Tools must be an array-like JsonContainer.");
    }
    m_tools = tools;
    return *this;
}

const JsonContainer& ChatHistory::get_tools() const {
    return m_tools;
}

ChatHistory& ChatHistory::set_extra_context(const JsonContainer& extra_context) {
    if (!extra_context.is_object()) {
        OPENVINO_THROW("Extra context must be an object-like JsonContainer.");
    }
    m_extra_context = extra_context;
    return *this;
}

const JsonContainer& ChatHistory::get_extra_context() const {
    return m_extra_context;
}

} // namespace genai
} // namespace ov

File: src/js/lib/chatHistory.ts
/* eslint-disable @typescript-eslint/no-misused-new */
// Copyright (C) 2025 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

/**
 * Represents a single message in a chat conversation.
 */
export type ChatMessage = Record<string, any>;

/**
 * Represents a tool definition for function calling.
 */
export type ToolDefinition = Record<string, any>;

/**
 * Extra context for custom template variables.
 */
export type ExtraContext = Record<string, any>;

/**
 * ChatHistory manages conversation messages and optional metadata for chat templates.
 *
 * @example
 * ```typescript
 * // Create an empty chat history
 * const history = new ChatHistory();
 *
 * // Add messages
 * history.push({ role: "system", content: "You are a helpful assistant." });
 * history.push({ role: "user", content: "Hello!" });
 * history.push({ role: "assistant", content: "Hi! How can I help you?" });
 *
 * // Create with initial messages
 * const history2 = new ChatHistory([
 *   { role: "system", content: "You are a helpful assistant." },
 *   { role: "user", content: "What's the weather?" }
 * ]);
 *
 * // Set tools for function calling
 * history.setTools([
 *   {
 *     type: "function",
 *     function: {
 *       name: "get_weather",
 *       description: "Get the current weather",
 *       parameters: {
 *         type: "object",
 *         properties: {
 *           location: { type: "string" }
 *         }
 *       }
 *     }
 *   }
 * ]);
 * ```
 */
export interface ChatHistory {
  /**
   * Creates a new ChatHistory instance.
   * @param messages - Optional array of initial messages
   */
  new (messages?: ChatMessage[]): ChatHistory;

  /**
   * Adds a message to the end of the chat history.
   * @param message - The message to add
   * @returns The ChatHistory instance for method chaining
   *
   * @example
   * ```typescript
   * history.push({ role: "user", content: "Hello!" });
   * ```
   *
   * @note The message structure is flexible and depends on the model used.
   */
  push(message: ChatMessage): ChatHistory;
  /**
   * Removes the last message from the chat history.
   * @throws {Error} If the history is empty
   */
  pop(): void;
  /**
   * Returns all messages in the chat history.
   * @returns Array of messages
   */
  getMessages(): ChatMessage[];
  /**
   * Replace all messages with a new list.
   *
   * @param messages - The new list of messages
   * @returns The ChatHistory instance for method chaining
   */
  setMessages(messages: ChatMessage[]): ChatHistory;
  /**
   * Removes all messages from the chat history.
   */
  clear(): void;
  /**
   * Returns the number of messages in the chat history.
   * @returns The number of messages
   */
  size(): number;
  /**
   * Checks if the chat history is empty.
   * @returns true if the history contains no messages, false otherwise
   *
   * @example
   * ```typescript
   * if (history.empty()) {
   *   console.log("No messages yet");
   * }
   * ```
   */
  empty(): boolean;
  /**
   * Sets the tools array for function calling.
   * @param tools - Array of tool definitions
   * @returns The ChatHistory instance for method chaining
   *
   * @example
   * ```typescript
   * history.setTools([
   *   {
   *     type: "function",
   *     function: {
   *       name: "get_weather",
   *       description: "Get current weather",
   *       parameters: {
   *         type: "object",
   *         properties: {
   *           location: { type: "string", description: "City name" }
   *         },
   *         required: ["location"]
   *       }
   *     }
   *   }
   * ]);
   * ```
   *
   * @note The tool structure is flexible and depends on the model used.
   */
  setTools(tools: ToolDefinition[]): ChatHistory;
  /**
   * Gets the tools array.
   * @returns The tools array
   */
  getTools(): ToolDefinition[];
  /**
   * Sets extra context for custom template variables.
   * @param context - Object containing custom variables
   * @returns The ChatHistory instance for method chaining
   *
   * @example
   * ```typescript
   * history.setExtraContext({
   *   user_name: "Alice",
   *   date: "2025-01-01",
   *   custom_instruction: "Be concise"
   * });
   * ```
   *
   * @note The extra context structure is flexible and depends on the model used.
   */
  setExtraContext(context: ExtraContext): ChatHistory;
  /**
   * Gets the extra context.
   * @returns The extra context object
   */
  getExtraContext(): ExtraContext;
}
